#LAB4
##OPTIMIZATION IN MACHINE LEARNING
###kshitij jaiswal (b20cs028)

In [ ]:
import numpy as np
from cvxopt import matrix, solvers

In [ ]:
#1
A = np.array([[1, 2], [-4, -3], [3, 1], [-3, -1], [1, 0], [-1, 0], [0, -1]])
b1 = np.array([[3], [-6], [3], [-3], [1], [-15/17], [0]])
c1 = np.array([5, 2])

b2 = np.array([[3], [-6], [3], [-3], [15/17], [0], [0]])
c2 = np.array([3, 7])

soln1 = solvers.lp(matrix(c1, tc='d'), matrix(A, tc='d'), matrix(b1, tc='d'))
soln2 = solvers.lp(matrix(c2, tc='d'), matrix(A, tc='d'), matrix(b2, tc='d'))

# print(np.round_(soln1['x'], decimals=5), soln1['primal objective']) --> gives unbounded solution...
print(np.round_(soln2['x'], decimals=5), soln2['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  5.7047e+00  9.9982e+00  2e+01  4e-01  1e+00  1e+00
 1:  5.5660e+00  6.4286e+00  1e+00  7e-02  2e-01  3e-01
 2: -2.9860e+00  5.0740e+01  6e+03  2e+00  5e+00  4e+01
 3:  4.4972e+00  9.8214e+01  3e+03  2e-01  6e-01  9e+01
 4:  4.4985e+00  9.1329e+03  3e+05  2e-01  6e-01  9e+03
 5:  4.4985e+00  9.1261e+05  3e+07  2e-01  6e-01  9e+05
 6:  4.4985e+00  9.1260e+07  3e+09  2e-01  6e-01  9e+07
Certificate of primal infeasibility found.
     pcost       dcost       gap    pres   dres   k/t
 0:  9.1085e+00  1.4601e+01  3e+01  4e-01  1e+00  1e+00
 1:  9.1442e+00  1.0121e+01  3e+00  6e-02  2e-01  4e-01
 2:  1.0175e+01  1.0218e+01  1e-01  3e-03  8e-03  1e-02
 3:  1.0200e+01  1.0200e+01  1e-03  3e-05  8e-05  1e-04
 4:  1.0200e+01  1.0200e+01  1e-05  3e-07  8e-07  1e-06
 5:  1.0200e+01  1.0200e+01  1e-07  3e-09  8e-09  1e-08
Optimal solution found.
[[0.6]
 [1.2]] 10.199999974652686


Optimal solutions occurs at (x1, x2) = (0,6, 1.2) and the optimum value of Z is 10.2

In [ ]:
#2
n = 3
m = 5

A = np.zeros(((n<<1)+m+(n*m), n*m))

for i in range(0, n<<1, 2):
    for j in range((i>>1)*m, ((i>>1)+1)*m):
        # print(i, j)
        A[i][j] = 1
        A[i+1][j] = -1


for i in range(n<<1, (n<<1)+m):
    for j in range(0, n*m, m):
        A[i,j + (i-(n<<1))%m] = 1


for i in range((n<<1)+m, (n<<1)+m+(n*m)):
    A[i, i-((n<<1)+m)]=-1

# for i in range((n<<1)+m+(n*m)):
#     for j in range(n*m):
#         print(int(A[i, j]), end=" ")
#     print("\n")

# print(A)

b = np.array([[8], [-8], [12], [-12],[14], [-14], [7], [5], [6], [8], [8]])
b = np.row_stack((b, np.zeros((n*m, 1))))
# print(b)

c = np.array([[4], [2], [3], [2], [6], [5], [4], [5], [2], [1], [6], [5], [4], [7], [7]])

# # print(A)
# # print(B)

soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))
print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  1.3932e+02  1.9545e+01  3e+02  3e-01  2e-16  1e+00
 1:  1.1599e+02  7.3132e+01  1e+02  1e-01  7e-16  2e+00
 2:  1.0596e+02  1.0016e+02  1e+01  2e-02  4e-15  3e-01
 3:  1.0314e+02  1.0238e+02  2e+00  2e-03  4e-15  5e-02
 4:  1.0300e+02  1.0299e+02  3e-02  4e-05  5e-16  9e-04
 5:  1.0300e+02  1.0300e+02  3e-04  4e-07  2e-15  9e-06
 6:  1.0300e+02  1.0300e+02  3e-06  4e-09  2e-15  9e-08
Optimal solution found.
[[0.]
 [4.]
 [0.]
 [4.]
 [0.]
 [0.]
 [0.]
 [0.]
 [4.]
 [8.]
 [7.]
 [1.]
 [6.]
 [0.]
 [0.]] 103.00000034914879


Transportations that occurs in the optimal case --
1. D1 --> S2 - 4 units
2. D1 --> S4 - 4 units
3. D2 --> S4 - 4 units
4. D2 --> S5 - 8 units
5. D3 --> S1 - 7 units
6. D3 --> S2 - 1 unit
7. D3 --> S3 - 6 units
Minimum Cost of above transportations - 103

In [ ]:
#3
A = np.array([[1, 1, 1, 0, 0, 0, 0, 0, 0], [-1, -1, -1, 0, 0, 0, 0, 0, 0], [-1, 0, 0, 1, 1, 0, 0, 0, 0],
              [1, 0, 0, -1, -1, 0, 0, 0, 0], [0, -1, 0, -1, 0, 1, 1, 0, 0], [0, 1, 0, 1, 0, -1, -1, 0, 0],
              [0, 0, 0, 0, -1, -1, 0, 1, 0], [0, 0, 0, 0, 1, 1, 0, -1, 0], [0, 0, -1, 0, 0, 0, -1, 0, 1],
              [0, 0, 1, 0, 0, 0, 1, 0, -1], [0, 0, 0, 0, 0, 0, 0, -1, -1], [0, 0, 0, 0, 0, 0, 0, 1, 1],
              [-1, 0, 0, 0, 0, 0, 0, 0, 0], [0, -1, 0, 0, 0, 0, 0, 0, 0], [0, 0, -1, 0, 0, 0, 0, 0, 0],
              [0, 0, 0, -1, 0, 0, 0, 0, 0], [0, 0, 0, 0, -1, 0, 0, 0, 0], [0, 0, 0, 0, 0, -1, 0, 0, 0],
              [0, 0, 0, 0, 0, 0, -1, 0, 0], [0, 0, 0, 0, 0, 0, 0, -1, 0], [0, 0, 0, 0, 0, 0, 0, 0, -1]])
b = np.array([[1], [-1], [3], [-3], [0], [0], [0], [0], [0], [0], [-4], [4], [0], [0], [0], [0], [0], [0], [0], [0], [0]])
c = np.array([[2], [3], [3], [2], [4], [1], [2], [3], [1]])

soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))

print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  1.6755e+01  1.6755e+01  1e+02  1e+00  9e-01  1e+00
 1:  1.9879e+01  2.0379e+01  1e+01  2e-01  1e-01  7e-01
 2:  1.9159e+01  1.9241e+01  2e+00  3e-02  2e-02  1e-01
 3:  1.9002e+01  1.9003e+01  2e-02  3e-04  3e-04  1e-03
 4:  1.9000e+01  1.9000e+01  2e-04  3e-06  3e-06  1e-05
 5:  1.9000e+01  1.9000e+01  2e-06  3e-08  3e-08  1e-07
Optimal solution found.
[[-0.]
 [ 0.]
 [ 1.]
 [ 3.]
 [-0.]
 [ 0.]
 [ 3.]
 [ 0.]
 [ 4.]] 19.000000160934945


Initially there was 1 wagon at Node 1 and 3 wagons at Node 2.
Optimum Path for wagon transportations -
1. wagon at Node 1 follows the path 1 --> 3 --> 5 --> 6.
2. All 3 wagons at Node 2 follow the path 2 --> 3 --> 5 --> 6.

Optimum Cost for transportations of wagon is 19.

In [ ]:
#4
A = np.array([[1, 1, 0, 0, 0], [-1, -1, 0, 0, 0], [-1, 0, 1, 1, 0], [1, 0, -1, -1, 0], [0, -1, -1, 0, 1],
              [0, 1, 1, 0, -1], [0, 0, 0, -1, -1], [0, 0, 0, 1, 1], [1, 0, 0, 0, 0], [0, 1, 0, 0, 0],
              [0, 0, 1, 0, 0], [0, 0, 0, 1, 0], [0, 0, 0, 0, 1], [-1, 0, 0, 0, 0], [0, -1, 0, 0, 0],
              [0, 0, -1, 0, 0], [0, 0, 0, -1, 0], [0, 0, 0, 0, -1]])
b = np.array([[2], [-2], [2], [-2], [-2], [2], [-2], [2], [1], [3], [2], [2], [3], [0], [0], [0], [0], [0]])
c = np.array([[1], [5], [1], [4], [2]])

soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))

print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  1.6200e+01 -6.9000e+00  7e+01  7e-01  3e-16  1e+00
 1:  1.3537e+01  1.1368e+01  5e+00  7e-02  5e-16  4e-01
 2:  1.3984e+01  1.3810e+01  4e-01  6e-03  1e-15  3e-02
 3:  1.4000e+01  1.3997e+01  5e-03  8e-05  2e-15  4e-04
 4:  1.4000e+01  1.4000e+01  5e-05  8e-07  4e-15  4e-06
 5:  1.4000e+01  1.4000e+01  5e-07  8e-09  3e-15  4e-08
Optimal solution found.
[[1.]
 [1.]
 [2.]
 [1.]
 [1.]] 13.999999958536225


Water that flows for each set of Nodes --
1. 1 --> 2 - 1 unit
2. 1 --> 3 - 1 unit
3. 2 --> 3 - 2 units
4. 2 --> 4 - 1 unit
5. 3 --> 4 - 1 unit

Optimum cost for the above water flow is 14.

In [ ]:
#5
class node:

    def __init__(self, val, edges) -> None:
        self.val = val
        self.incoming = []
        self.outgoing = []
        for (u, v) in edges:
            if self.val == u:
                self.outgoing.append(v)
            if self.val == v:
                self.incoming.append(u)

if __name__ == '__main__':

    numNodes = 10
    numEdges = 20

    start = 1
    end = numNodes

    edges = [(1, 2), (1, 3), (1, 4), (2, 5), (2, 6), (2, 7), (3, 5), (3, 6), (3, 7), (4, 5), (4, 6), (4, 7), (5, 8), (5, 9), (6, 8), (6, 9), (7, 8), (7, 9), (8, 10), (9, 10)]

    lstNodes = []

    for i in range(1, numNodes+1):
        x = node(i, edges)
        lstNodes.append(x)

    A = np.zeros(((numNodes + numEdges)<<1, numEdges))

    for i in range(0, numNodes<<1, 2):
        obj = lstNodes[i>>1]
        val = obj.val
        for out in obj.outgoing:
            A[i, edges.index((val, out))] = 1
            A[i+1, edges.index((val, out))] = -1
        for inc in obj.incoming:
            A[i, edges.index((inc, val))] = -1
            A[i+1, edges.index((inc, val))] = 1

    for i in range(numNodes<<1, (numNodes<<1)+numEdges):
        A[i, i-(numNodes<<1)] = -1
        A[i+numEdges, i-(numNodes<<1)] = 1

    # print(A)

    b = np.zeros(((numNodes + numEdges)<<1, 1))
    for i in range(0, numNodes<<1, 2):
        nodeNum = (i>>1) + 1
        if nodeNum == start:
            b[i] = 1
            b[i+1] = -1
        elif nodeNum == end:
            b[i] = -1
            b[i+1] = 1
        else:
            b[i] = 0
            b[i+1] = 0

    for i in range((numNodes<<1)+numEdges, (numNodes + numEdges)<<1):
        b[i] = 1

    # print(b)
    c = np.array([[4], [6], [6], [6], [8], [9], [5], [4], [6], [5], [5], [7], [6], [8], [4], [9], [3], [7], [9], [6]])

    soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))
    print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  3.8337e+01 -5.8837e+01  5e+02  2e+00  3e-16  1e+00
 1:  1.9212e+01  8.0898e+00  3e+01  2e-01  3e-16  6e-01
 2:  2.3319e+01  2.0775e+01  6e+00  6e-02  3e-15  2e-01
 3:  2.3010e+01  2.2468e+01  1e+00  1e-02  5e-15  3e-02
 4:  2.3005e+01  2.2989e+01  4e-02  4e-04  1e-14  9e-04
 5:  2.3000e+01  2.3000e+01  4e-04  4e-06  5e-14  9e-06
 6:  2.3000e+01  2.3000e+01  4e-06  4e-08  3e-14  9e-08
Optimal solution found.
[[ 0.]
 [ 1.]
 [ 0.]
 [ 0.]
 [-0.]
 [ 0.]
 [-0.]
 [ 1.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 1.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 1.]
 [ 0.]] 23.000000521623715


Shortest path from 1 --> 10 is 1 --> 3 --> 6 --> 8 --> 10. Hence the minimum cost is 6 + 4 + 4 + 9 = 23

In [ ]:
#6
n = 4

A = np.zeros(((n<<2) + n*n, n*n))

for i in range(0, n<<1, 2):
    for j in range((i>>1)*n, ((i>>1)+1)*n):
        # print(i, j)
        A[i][j] = 1
        A[i+1][j] = -1


for i in range(n<<1, n<<2, 2):
    for j in range(0, n*n, n):
        # print(i, j + (i>>1)%n)
        A[i,j + (i>>1)%n] = 1
        A[i+1,j + (i>>1)%n] = -1


for i in range(n<<2, (n<<2) + n*n, 1):
    A[i, i-(n<<2)]=-1

# for i in range((n<<2)+(n*n)):
#     for j in range(n*n):
#         print(int(A[i, j]), end=" ")
#     print("\n")

b = np.zeros(((n<<2) + n*n, 1))
for i in range(n<<2):
    if i&1:
        b[i, 0] = -1
    else:
        b[i, 0] = 1

c = np.array([[20], [28], [19], [13], [15], [30], [31], [28], [40], [21], [20], [17], [21], [28], [26], [12]])

# print(A)
# print(b)

soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))
print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  8.6824e+01  8.6824e+01  4e+02  1e+00  5e-01  1e+00
 1:  6.6618e+01  6.6857e+01  3e+01  1e-01  4e-02  3e-01
 2:  6.7230e+01  6.7297e+01  4e+00  2e-02  6e-03  8e-02
 3:  6.7002e+01  6.7003e+01  4e-02  2e-04  7e-05  9e-04
 4:  6.7000e+01  6.7000e+01  4e-04  2e-06  7e-07  9e-06
 5:  6.7000e+01  6.7000e+01  4e-06  2e-08  7e-09  9e-08
Optimal solution found.
[[ 0.]
 [ 0.]
 [ 1.]
 [ 0.]
 [ 1.]
 [ 0.]
 [ 0.]
 [-0.]
 [-0.]
 [ 1.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 1.]] 67.00000017775807


Optimal distribution of Tasks (A - D) among Operators (1 - 4) is --
3 does A
1 does B
2 does C
4 does D
Hence the minimum cost of operating all tasks is 67.


In [ ]:
#7
n = 5
A = np.zeros(((n<<2) + n*n, n*n))

for i in range(0, n<<1, 2):
    for j in range((i>>1)*n, ((i>>1)+1)*n):
        # print(i, j)
        A[i][j] = 1
        A[i+1][j] = -1

for i in range(n<<1, n<<2, 2):
    for j in range(0, n*n, n):
        # print(i, j + (i>>1)%n)
        A[i,j + (i>>1)%n] = 1
        A[i+1,j + (i>>1)%n] = -1

for i in range(n<<2, (n<<2) + n*n, 1):
    A[i, i-(n<<2)]=-1

# for i in range((n<<2)+(n*n)):
#     for j in range(n*n):
#         print(int(A[i, j]), end=" ")
#     print("\n")

b = np.zeros(((n<<2) + n*n, 1))
for i in range(n<<2):
    if i&1:
        b[i, 0] = -1
    else:
        b[i, 0] = 1

c = np.array([[37.7], [32.9], [33.8], [37.0], [35.4], [43.4], [33.1], [42.2], [34.7], [41.8], [33.3], [28.5], [38.9], [30.4], [33.6], [29.2], [26.4], [29.6], [28.5], [31.1]])
c = np.row_stack((c, np.zeros((n, 1))))
# print(A)
# print(b)

# print(c)

soln = solvers.lp(matrix(c, tc='d'), matrix(A, tc='d'), matrix(b, tc='d'))
print(np.round_(soln['x'], decimals=5), soln['primal objective'])

     pcost       dcost       gap    pres   dres   k/t
 0:  1.2981e+02  1.2981e+02  7e+02  2e+00  4e-01  1e+00
 1:  1.1706e+02  1.1759e+02  5e+01  2e-01  5e-02  7e-01
 2:  1.2660e+02  1.2679e+02  7e+00  4e-02  1e-02  2e-01
 3:  1.2600e+02  1.2605e+02  2e+00  9e-03  2e-03  5e-02
 4:  1.2620e+02  1.2620e+02  1e-01  6e-04  2e-04  4e-03
 5:  1.2620e+02  1.2620e+02  1e-03  6e-06  2e-06  4e-05
 6:  1.2620e+02  1.2620e+02  1e-05  6e-08  2e-08  4e-07
Optimal solution found.
[[0.     ]
 [0.     ]
 [1.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.99999]
 [0.     ]
 [0.     ]
 [0.99999]
 [0.     ]
 [0.     ]
 [0.     ]
 [1.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [0.     ]
 [1.     ]] 126.19999967106038


Optimal Distribution of Swimmers across all Strokes is -->
1. David swims Backstroke. (33.8)
2. Tony swims Breaststroke. (34.7)
3. Chris swims Butterfly. (28.5)
4. Carl swims Freestyle. (29.2)

Ken DOES NOT swim.

Minimum achievable time for all strokes combined is 33.8 + 34.7 + 28.5 + 29.2 = 126.2 seconds.

